contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+
+JSON-Schema also allows us to reject additional fields that are not defined in the schema and only allow the fields that are specified. This can be achieved by using `"additionalProperties": false` in the schema. However, there is danger in allowing CRD authors to set `"additionalProperties": false` because it breaks version skew (new client can send new optional fields to the old server). So we should not allow CRD authors to set `"additionalProperties": false`.
+
+### Client-Side Validation
+
+The client-side validation is carried out before sending the request to the api-server, or even completely offline. This can be achieved while creating resources through the client i.e. kubectl using the --validate option.
+
+If the API type serves the JSON-Schema in the swagger spec, the existing kubectl code will already be able to also validate CRs. This will be achieved as a follow-up.
+
+### Comparison between server-side and client-side Validation
+
+The table below shows the cases when server-side and client-side validation methods are applicable.
+
+| Case                                               | Server-Side  | Client-Side   |
+|:--------------------------------------------------:|:------------:|:-------------:|
+| Kubectl create/edit/replace with validity feedback | &#x2713;     | &#x2713;      | 
+| Custom controller creates/updates CRs              | &#x2713;     | &#x2717;      |
+| CRs are created by an untrusted party              | &#x2713;     | &#x2717;      |
+| Not making validation for CRs a special case       | &#x2713;     | &#x2717;      |
+
+The above table is an evidence that we need server-side validation as well, next to the client-side validation we easily get, nearly for free, by serving Swagger/OpenAPI specs in apiextension-apiserver. 
+
+This is especially true in situations when CRs are used by components that are out of the control of the admin. Example: A user can create a database CR for a Database-As-A-Service. In this case, only server-side validation can give confidence that the CRs are well formed.
+
+### Existing Instances and changing the Schema
+
+If the schema is made stricter later, the existing CustomResources might no longer comply with the spec. This will make them unchangeable and essentially read-only.
+
+To avoid this, it is the responsibility of the user to make sure that any changes made to the schema are such that the existing CustomResources remain validated. 
+ 
+Note:
+
+1. This is the same behavior that we require for native resources. Validation cannot be made stricter in later Kubernetes versions without breaking compatibility.
+
+2. For migration of CRDs with no validation to CRDs with validation, we can create a controller that will validate and annotate invalid CRs once the spec changes, so that the custom controller can choose to delete them (this is also essentially the status condition of the CRD). This can be achieved, but it is not part of the proposal.
```

Comment:
I'm not really a fan of this.  It seems like it should mirror kube-apis: it's the responsibility of the APi author and fixups are performed client-side.